indexing-service/src/main/java/org/apache/druid/indexing/overlord/supervisor/SupervisorResource.java

```diff
@@ -132,24 +137,39 @@ public Response specGetAll(
           );
           final boolean includeFull = full != null;
           final boolean includeState = state != null && state;
+          final boolean includeFullStatus = fullStatus != null;
 
-          if (includeFull || includeState) {
-            List<Map<String, ?>> allStates = authorizedSupervisorIds
+          if (includeFull || includeState || includeFullStatus) {
+            List<SupervisorStatus> allStates = authorizedSupervisorIds
                 .stream()
                 .map(x -> {
                   Optional<SupervisorStateManager.State> theState =
                       manager.getSupervisorState(x);
-                  ImmutableMap.Builder<String, Object> theBuilder = ImmutableMap.builder();
-                  theBuilder.put("id", x);
+                  SupervisorStatus.Builder theBuilder = new SupervisorStatus.Builder();
+                  theBuilder.withId(x);
                   if (theState.isPresent()) {
-                    theBuilder.put("state", theState.get().getBasicState());
-                    theBuilder.put("detailedState", theState.get());
-                    theBuilder.put("healthy", theState.get().isHealthy());
+                    theBuilder.withState(theState.get().getBasicState().toString())
+                              .withDetailedState(theState.get().toString())
+                              .withHealthy(theState.get().isHealthy());
                   }
                   if (includeFull) {
                     Optional<SupervisorSpec> theSpec = manager.getSupervisorSpec(x);
                     if (theSpec.isPresent()) {
-                      theBuilder.put("spec", theSpec.get());
+                      theBuilder.withSpec(manager.getSupervisorSpec(x).get());
+                    }
+                  }
+                  if (includeFullStatus) {
+                    Optional<SupervisorSpec> theSpec = manager.getSupervisorSpec(x);
+                    if (theSpec.isPresent()) {
+                      try {
+                        theBuilder.withSpecString(objectMapper.writeValueAsString(manager.getSupervisorSpec(x).get()));
```

Comment:
It may be nice to have to builder accept SupervisorSpec and serialize/deserialize it to a string internally so that the formatting logic is all in one place.

Reply:
i tried that, but the `objectMapper` is injected here, which is used to serialize the SupervisorSpec

Reply:
Does it work if you add another constructor for `SupervisorStatus.Builder` that injects the `ObjectMapper`?

Reply:
hmm, it seems odd to add `ObjectMapper` to Builder, and the builder already accepts a `spec`,  in addition to `specString`. I could generate `specString` from `spec` in `SupervisorStatus#getSpecString` if I have the right `ObjectMapper`, but then I think the `specString` would appear in response to `druid/indexer/v1/supervisor?full` as well as `druid/indexer/v1/supervisor?fullStatus`, since it's not null anymore for the former call and it's not required/desired in former call.

Reply:
Oh, since the SupervisorStatus DTO already has a SupervisorSpec field that gets serialized when the response is serialized, why do you need to have another field that's an explicitly serialized version (which is populated before the response is serialized)?

Reply:
Yeah, i added already serialized `specString` to get the json payload in `SystemSchema#SupervisorsTable` via the `JsonParserIterator`, otherwise the deserialization errors out since `KafkaSupervisorSpec`'s attributes are not present in broker, they are only available in overlord.

Reply:
@ccaominh added comment and javadoc about the need for explicitly serialized `spec`, let me know if it helps clarify things